### ISSUE TITLE
dynamodb-local inside Gitlab CI

### ISSUE THREAD

**Issue by Meemaw** [testcontainers/testcontainers-node#7]
I can't get [amazon/dynamodb-local](https://hub.docker.com/r/amazon/dynamodb-local/) image work inside Gitlab CI. Note that it works locally without any problems. I haven't tried if this works in [testcontainers-java](https://github.com/testcontainers/testcontainers-java), cc @rnorth

I'm using CI settings from [docs](https://www.testcontainers.org/supported_docker_environment/continuous_integration/gitlab_ci/).

My `.gitlab-ci.yml`:

```yml
image: node:12

stages:
  - prepare
  - test

# DinD service is required for Testcontainers
services:
  - docker:dind

variables:
  DOCKER_HOST: 'tcp://docker:2375'
  DOCKER_DRIVER: overlay2

install_packages_run_linters:
  stage: prepare
  script:
    - yarn install
    - yarn lint
  artifacts:
    expire_in: 1 hour
    paths:
      - ./node_modules

test:unit:
  stage: test
  script:
    - DEBUG=testcontainers yarn test --runInBand --coverage
  coverage: '/All files[^|]*\|[^|]*\s+([\d\.]+)/'
  dependencies:
    - install_packages_run_linters
```

I'm setting up container like this:

```typescript
const startedContainer = new GenericContainer("amazon/dynamodb-local").withExposedPorts(8000).start()
```

Relevant CI output:

````
[0KRunning with gitlab-runner 11.11.0-rc2 (7f58b1ec)
[0;m[0K  on docker-auto-scale 0277ea0f
[0;msection_start:1557296045:prepare_executor
[0K[0KUsing Docker executor with image node:12 ...
[0;m[0KStarting service docker:dind ...
[0;m[0KPulling docker image docker:dind ...
[0;m[0KUsing docker image sha256:6204caf6c5927058f10fab0e4661b5f7a37d9a986a89b28a176a9958a0a5b7d3 for docker:dind ...
[0;m[0KWaiting for services to be up and running...
[0;m[0KPulling docker image node:12 ...
[0;m[0KUsing docker image sha256:c77f0d290562eefd2c9e5481cc1707f40a43d2dd5f7719384bd3b8f05699ffd7 for node:12 ...
[0;msection_end:1557296093:prepare_executor
[0Ksection_start:1557296093:prepare_script
[0KRunning on runner-0277ea0f-project-12215658-concurrent-0 via runner-0277ea0f-srm-1557295996-5afcffc7...
section_end:1557296098:prepare_script
[0Ksection_start:1557296098:get_sources
[0KInitialized empty Git repository in /builds/Meemaw/meshwatch-backend-core/.git/
[32;1mFetching changes...[0;m
[32;1mCreated fresh repository.[0;m
From https://gitlab.com/Meemaw/meshwatch-backend-core
 * [new branch]      master     -> origin/master
[32;1mChecking out fb1e6645 as master...[0;m

[32;1mSkipping Git submodules setup[0;m
section_end:1557296100:get_sources
[0Ksection_start:1557296100:restore_cache
[0Ksection_end:1557296101:restore_cache
[0Ksection_start:1557296101:download_artifacts
[0K[32;1mDownloading artifacts for install_packages_run_linters (208940828)...[0;m
Downloading artifacts from coordinator... ok      [0;m  id[0;m=208940828 responseStatus[0;m=200 OK token[0;m=oi_pU5bU
section_end:1557296112:download_artifacts
[0Ksection_start:1557296112:build_script
[0K[32;1m$ DEBUG=testcontainers yarn test --runInBand --coverage[0;m
yarn run v1.15.2
warning package.json: No license field
$ tsdx test --verbose --runInBand --coverage
2019-05-08T06:15:26.975Z testcontainers Pulling image: amazon/dynamodb-local:latest
2019-05-08T06:15:44.045Z testcontainers Creating container for image: amazon/dynamodb-local:latest
2019-05-08T06:15:44.875Z testcontainers Starting container with ID: 32d5a7fed1d39799e6093fa46c52e8901156043e3b1bdf2bedd937d486133ff9
2019-05-08T06:15:46.372Z testcontainers Waiting for host port :39113
2019-05-08T06:15:46.376Z testcontainers Waiting for internal port :8000
FAIL test/monitoring-service.spec.ts (72.17s)
  MonitoringService
    with DynamoStorage
      bookmarkMonitor
        âœ• should_bookmarkMonitor_when_monitorIdPassed (14ms)
        âœ• should_return404_when_bookmarkMonitorWithRandomId
      deleteMonitor
        âœ• should_deleteMonitor__when_newlyCreatedMonitorParamsPassed (1ms)
        âœ• should_return404_when_deleteMonitorWithRandomId
      getMonitorsByScheduler
        âœ• should_getEmptyCollection_when_searchByRandomScheduler (1ms)
        âœ• should_getMonitors_when_searchByTestScheduler
      getMonitors
        âœ• should_getEmptyCollection_when_searchByRandomUserId
        âœ• should_getTestMonitors_when_searchByTestUserId (1ms)
      createMonitor
        âœ• should_createNewMonitor_when_createMonitorPayloadPassed
        âœ• should_returnBoomWithFieldError_when_invalidCreateMonitorPayloadPassed
        âœ• should_returnBoomWithFieldErrors_when_emptyObjectPassed (1ms)
      updateMonitor
        âœ• should_updateMonitor_when_updateMonitorPayloadIsPassed
        âœ• should_returnBoomWithFieldError_when_createMonitorPayloadPassed (1ms)

  â— MonitoringService â€º with DynamoStorage â€º bookmarkMonitor â€º should_bookmarkMonitor_when_monitorIdPassed

    Port :39113 not bound after 40000ms
```

**Comment by cristianrgreco** [comment]
Hi @Meemaw, it seems from the logs that the container is unable to bind its ports. Interesting that it works locally and not in GitLabs as the environment *shouldn't* matter here.

Version `1.1.12` fixes issue https://github.com/testcontainers/testcontainers-node/issues/5 which changes the behaviour of internal port checking.
Could you please upgrade, try again and report back?

**Comment by Meemaw** [comment]
Upgraded but the error persists:

CI output:

```
Running with gitlab-runner 11.11.0-rc2 (7f58b1ec)
  on docker-auto-scale fa6cab46
Using Docker executor with image node:12 ...
Starting service docker:dind ...
Pulling docker image docker:dind ...
Using docker image sha256:6204caf6c5927058f10fab0e4661b5f7a37d9a986a89b28a176a9958a0a5b7d3 for docker:dind ...
Waiting for services to be up and running...
Pulling docker image node:12 ...
Using docker image sha256:c77f0d290562eefd2c9e5481cc1707f40a43d2dd5f7719384bd3b8f05699ffd7 for node:12 ...
Running on runner-fa6cab46-project-12215658-concurrent-0 via runner-fa6cab46-srm-1557305768-55e74cdb...
Initialized empty Git repository in /builds/Meemaw/meshwatch-backend-core/.git/
Fetching changes...
Created fresh repository.
From https://gitlab.com/Meemaw/meshwatch-backend-core
 * [new branch]      master     -> origin/master
Checking out 0ce92027 as master...

Skipping Git submodules setup
Downloading artifacts for install_packages_run_linters (209030943)...
Downloading artifacts from coordinator... ok        id=209030943 responseStatus=200 OK token=nzzzxfzs
$ DEBUG=testcontainers yarn test --runInBand --coverage
yarn run v1.15.2
warning package.json: No license field
$ tsdx test --verbose --runInBand --coverage
2019-05-08T08:58:20.699Z testcontainers Pulling image: amazon/dynamodb-local:latest
2019-05-08T08:58:37.674Z testcontainers Creating container for image: amazon/dynamodb-local:latest
2019-05-08T08:58:38.260Z testcontainers Starting container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:39.871Z testcontainers Waiting for host port :40621
2019-05-08T08:58:39.875Z testcontainers Waiting for internal port :8000
2019-05-08T08:58:39.875Z testcontainers Executing command "/bin/sh -c cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :1f40" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:39.878Z testcontainers Executing command "/bin/sh -c nc -vz -w 1 localhost 8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:39.879Z testcontainers Executing command "/bin/bash -c </dev/tcp/localhost/8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:40.687Z testcontainers Executing command "/bin/sh -c cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :1f40" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:40.689Z testcontainers Executing command "/bin/sh -c nc -vz -w 1 localhost 8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:40.690Z testcontainers Executing command "/bin/bash -c </dev/tcp/localhost/8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:41.440Z testcontainers Executing command "/bin/sh -c cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :1f40" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:41.442Z testcontainers Executing command "/bin/sh -c nc -vz -w 1 localhost 8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:41.442Z testcontainers Executing command "/bin/bash -c </dev/tcp/localhost/8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:42.243Z testcontainers Executing command "/bin/sh -c cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :1f40" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:42.244Z testcontainers Executing command "/bin/sh -c nc -vz -w 1 localhost 8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:42.245Z testcontainers Executing command "/bin/bash -c </dev/tcp/localhost/8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:43.055Z testcontainers Executing command "/bin/sh -c cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :1f40" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:43.056Z testcontainers Executing command "/bin/sh -c nc -vz -w 1 localhost 8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:43.057Z testcontainers Executing command "/bin/bash -c </dev/tcp/localhost/8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:43.821Z testcontainers Executing command "/bin/sh -c cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :1f40" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:43.822Z testcontainers Executing command "/bin/sh -c nc -vz -w 1 localhost 8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:43.823Z testcontainers Executing command "/bin/bash -c </dev/tcp/localhost/8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:44.505Z testcontainers Executing command "/bin/sh -c cat /proc/net/tcp{,6} | awk '{print $2}' | grep -i :1f40" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:44.507Z testcontainers Executing command "/bin/sh -c nc -vz -w 1 localhost 8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
2019-05-08T08:58:44.507Z testcontainers Executing command "/bin/bash -c </dev/tcp/localhost/8000" on container with ID: b5926723b361a372a77fe574bf5d63a51b973275d0ca09a778bb8bd6b948416b
FAIL test/monitoring-service.spec.ts (72.263s)
  MonitoringService
    with DynamoStorage
      bookmarkMonitor
        ✕ should_bookmarkMonitor_when_monitorIdPassed (19ms)
        ✕ should_return404_when_bookmarkMonitorWithRandomId (1ms)
      deleteMonitor
        ✕ should_deleteMonitor__when_newlyCreatedMonitorParamsPassed (1ms)
        ✕ should_return404_when_deleteMonitorWithRandomId
      getMonitorsByScheduler
        ✕ should_getEmptyCollection_when_searchByRandomScheduler (1ms)
        ✕ should_getMonitors_when_searchByTestScheduler (8ms)
      getMonitors
        ✕ should_getEmptyCollection_when_searchByRandomUserId (8ms)
        ✕ should_getTestMonitors_when_searchByTestUserId (1ms)
      createMonitor
        ✕ should_createNewMonitor_when_createMonitorPayloadPassed (9ms)
        ✕ should_returnBoomWithFieldError_when_invalidCreateMonitorPayloadPassed (1ms)
        ✕ should_returnBoomWithFieldErrors_when_emptyObjectPassed (1ms)
      updateMonitor
        ✕ should_updateMonitor_when_updateMonitorPayloadIsPassed (6ms)
        ✕ should_returnBoomWithFieldError_when_createMonitorPayloadPassed (1ms)

  ● MonitoringService › with DynamoStorage › bookmarkMonitor › should_bookmarkMonitor_when_monitorIdPassed

    Port :40621 not bound after 40000ms
```

**Comment by cristianrgreco** [comment]
I can confirm that `amazon/dynamodb-local` works locally. 

I assume then that the issue is related to docker in docker, and that `testcontainers-node` is not using the correct socket in this case. I see you've set the `DOCKER_HOST` env variable, I'd need to check what `testcontainers-java` does with it. Possibly it's as simple as using that value as the socket instead of the default one if it is present.

**Comment by cristianrgreco** [comment]
@Meemaw Version `1.1.13` will now use the `DOCKER_HOST` env variable if one is set as the socket path. Could you please upgrade and try again?

**Comment by Meemaw** [comment]
Getting     `connect ENOENT tcp://docker:2375`

**Comment by cristianrgreco** [comment]
Ah my bad, it's not a socket. Will let you know once this is fixed.

**Comment by cristianrgreco** [comment]
@Meemaw Could you please try version `1.1.14`? 🙂

**Comment by Meemaw** [comment]
`Port :46427 not bound after 40000ms` again :|

**Comment by Meemaw** [comment]
Might be. Would be nice if this is fixed, as this is now unusable inside CI environments. If you need some help, I can fix/look into it once I'll find some time.

**Comment by cristianrgreco** [comment]
> Might be. Would be nice if this is fixed, as this is now unusable inside CI environments. If you need some help, I can fix/look into it once I'll find some time.

Help would be appreciated! 
The issue isn't CI environments, as the build pipeline runs in Travis - the issue is docker in docker.

Perhaps @rnorth could point me in the right direction?

**Comment by cristianrgreco** [comment]
Btw @Meemaw I noticed in your example that you're not `awaiting` the result of starting the container:
```
const startedContainer = new GenericContainer("amazon/dynamodb-local").withExposedPorts(8000).start()
```

**Comment by cristianrgreco** [comment]
@Meemaw I've verified that version `1.1.15` fixes this issue: https://gitlab.com/cristianrgreco/testcontainers-node/-/jobs/210895051

Could you please upgrade and try again?

**Comment by Meemaw** [comment]
I have upgraded and can confirm that he container now starts. However, there is still a problem. I am connecting to dynamodb locally like:

```javascript
const endpoint = `http://localhost:${startedContainer.getMappedPort(exposedPort)}`;
const dynamodb = new DynamoDB({ endpoint })
```

This doesn't work in CI, as the container is on different IP. It would be nice if testcontainers exposed a method like `getContainerIpAddress`. Java version already exposes [that](https://github.com/testcontainers/testcontainers-java/blob/2c5f274d73317257efa99d359357856c076c24b7/core/src/main/java/org/testcontainers/containers/ContainerState.java#L29).

**Comment by cristianrgreco** [comment]
@Meemaw This should be fixed now in version `1.1.17`, could you give it a go?

**Comment by Meemaw** [comment]
I can confirm that it works now 🚀 Thanks for fix and for being so responsive :+1:
